File: CWE121/C/stackBufferOverflow.c
/* To run native: 
    gcc stackBufferOverflow.c

   To run Web Assembly: 
    emcc -g stackBufferOverflow.c -o stackBufferOverflow.html
        1. In the js file enter your arguments in the "arguments_" array
        2. Launch local host server for the html file
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define BUFSIZE 16

void printStackData(char buf1[BUFSIZE], char buf2[BUFSIZE]) {
    printf("Stack Data 1: %s, Memory Address: %p\n", buf1, buf1);
    printf("Stack Data 2: %s, Memory Address: %p\n", buf2, buf2);
}

int main(int argc, char *argv[]) {
    
    printf("argv[1]: %s\n", argv[1]);

    char buf1[BUFSIZE], buf2[BUFSIZE];
    strcpy(buf1, "password");

    printf("\nExecute:\nchar buf1[%i], buf2[%i];\nstrcpy(buf1, \"password\");\n\n", BUFSIZE, BUFSIZE);

    printStackData(buf1, buf2);

    printf("\nExecute:\nstrcpy(buf2, argv[1]);\n\n");

    strcpy(buf2, argv[1]);
    
    printStackData(buf1, buf2);

    return 0;

}

File: CWE121/Go/stackBufferOverflow.go
/* To run native:
    go run stackBufferOverflow.go

   To run Web Assembly:
    GOOS=js GOARCH=wasm go build -o stackBufferOverflow.wasm stackBufferOverflow.go
        1. In the js file enter your arguments in the "this.argv" array on line 97
        2. Launch local host server for the html file
*/

package main

import (
	"fmt"
	"os"
)

const BUFSIZE = 16

func printStackData(buf1, buf2 []byte) {
	fmt.Printf("Stack Data 1: %s, Memory Address: %p\n", buf1, &buf1[0])
	fmt.Printf("Stack Data 2: %s, Memory Address: %p\n", buf2, &buf2[0])
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: program <input>")
		return
	}

	fmt.Printf("argv[1]: %s\n", os.Args[1])

	var buf1 [BUFSIZE]byte
	var buf2 [BUFSIZE]byte

	copy(buf1[:], "password")

	fmt.Printf("\nExecute:\nvar buf1 [%d]byte\nvar buf2 [%d]byte\ncopy(buf1[:], \"password\")\n\n", BUFSIZE, BUFSIZE)

	printStackData(buf1[:], buf2[:])

	fmt.Printf("\nExecute:\ncopy(buf2[:], os.Args[1])\n\n")

	copy(buf2[:], os.Args[1])

	printStackData(buf1[:], buf2[:])
}
